Stripe settings persistence - data layer

waclawjacek opened the pull request:
## Changes proposed in this Pull Request:

Fixes #1810.

This PR implements the API for saving settings, including:

* The `@wordpress/data` store, incl. hooks.
* The REST controller.
* The `Loadable` components.

I have only added saving support for the "Enable express checkouts" checkbox (excluding the location checkboxes) so there is a setting that can be tested but the PR is still focused mostly on the settings API. Support for saving other settings will be implemented in separate PRs.

![Markup on 2021-09-07 at 16:39:43](https://user-images.githubusercontent.com/1759681/132363882-1bb1900d-e0ec-4a3b-bf44-c5c2e8f32c5e.png)

## Testing instructions

1. Enable UPE preview, UPE, and UPE redesign.
1. Visit the settings page. Notice that as settings are loading:
   * A loading indicator box is displayed over the "Express checkouts" card.
   * The "Save changes" button is disabled.
1. Change the value of the "Enable express checkouts" checkbox.
1. Click "Save changes".
   * Notice the button is disabled and marked as busy while settings are saving.
   * Notice a snackbar notification informing of a successful save is displayed.
1. Refresh the page.
1. Notice that the checkbox value is persisted.
1. Toggle the checkbox, save, and refresh the page once more to confirm both the disabled/enabled status is saved.

## Some resolved struggles with compatibility checks

As part of this PR, due to the compatibility checks verifying compatibility with WC 3.3.0, WP 4.7, and PHP 5.6, I had to tweak some things to get the tests to pass. This includes:

* Deregistering WC Blocks' API endpoint (same reason as in WCPay).
  * I have tweaked this method though to do a `class_exists()` check to support even older versions than WCPay does.
* Removed type declarations from a method header (`bool`).
* Changed a test from `WP_REST_Request::has_param()` to `WP_REST_Request::get_param()` as `has_param` is not available in WP 4.7.
  * The only drawback is that now we can't distinguish between `null` and absent but this was not an issue for the code this PR adds.

Commit:
Add Settings REST controller with empty responses

Commit:
Add support for enabling/disabling payment requests in REST

Commit:
Add @wordpress/data settings store

Commit:
Add dependency on @wordpress/data and @wordpress/data-controls

Commit:
Add support for saving isPaymentRequestEnabled

Commit:
Port Loadable, LoadableBlock, and LoadableSettingsBlock

Commit:
Wrap PaymentRequestSection in LoadableSettingsSection

Commit:
Add client/data/settings tests

Commit:
PaymentRequestSection usePaymentRequestEnabledSettings() tests

Commit:
Add SaveSettingsSection data API tests

Commit:
WC_REST_Stripe_Settings_Controller tests

Commit:
Fix linting errors

Commit:
Merge branch 'develop' into add/1717-stripe-settings-persistence

Commit:
Remove extra bracket

Commit:
Require REST controllers in tests/phpunit/bootstrap.php

Commit:
Merge branch 'develop' into add/1717-stripe-settings-persistence

Commit:
Fix tests on older WC versions by deregistering WC Blocks endpoints

Also:

* Remove unused $controller and $gateway fields from test class.
* Revert requiring controller files in tests using rest_api_init

Commit:
Add extra check for WC 3.3.0 in test

Commit:
Update settings REST controller and tests for backward compatibility

frosso commented on client/settings/payment-request-section/__tests__/index.test.js:
```diff
@@ -3,9 +3,31 @@ import { screen, render } from '@testing-library/react';
+	usePaymentRequestEnabledSettings
+} from '../../../data/index';
+
+jest.mock( '../../../data/index', () => ( {
```
(minor) the `index` is not necessary

```suggestion
} from '../../../data';

jest.mock( '../../../data', () => ( {
```

frosso commented on client/settings/payment-request-section/__tests__/index.test.js:
```diff
@@ -3,9 +3,31 @@ import { screen, render } from '@testing-library/react';
+
 	it( 'should enable express checkout locations when express checkout is enabled', () => {
+		usePaymentRequestEnabledSettings.mockReturnValue(
+			getMockPaymentRequestEnabledSettings( true, jest.fn() )
```
(minor) what are your thoughts on just using the raw values for the tests (without the need for a `getMockPaymentRequestEnabledSettings` that is "building" an array with the values)?

```suggestion
		[ true, jest.fn() ]
```

waclawjacek replied:
Yes! :rocket:

frosso commented on woocommerce-gateway-stripe.php:
```diff
@@ -564,9 +564,16 @@ public function register_routes() {
 					$upe_flag_toggle_controller = new WC_REST_UPE_Flag_Toggle_Controller();
 					$upe_flag_toggle_controller->register_routes();
+
+					$gateway = new WC_Gateway_Stripe();
```
The Stripe gateway registers some filters/actions in its constructor. So instantiating it here means that the filters/actions are also registered (and executed, when called) multiple times.
Also, this is the `WC_Gateway_Stripe` being instantiated - but when the UPE flag is enabled it is the `WC_Stripe_UPE_Payment_Gateway` that is instantiated.

One possible solution could be to having a property on the `WC_Stripe` class, to ensure that the gateway is instantiated only once. This would also allow us to pass the most accurate gateway to the `WC_REST_Stripe_Settings_Controller` class.

waclawjacek replied:
Great catch! :100:

I tried going with the single-instance-stored-as-property approach but it turned out to be more complicated than expected. Just so we can move forward with this PR, I have used this instead:

```php
$stripe_gateway = WC()->payment_gateways()->payment_gateways()[ WC_Gateway_Stripe::ID ];
```

This guarantees a single instance without having to refactor any of the existing code.

frosso replied:
Great idea, since for both UPE and non-UPE checkout the gateway id is the same 👍

frosso commented on client/data/settings/hooks.js:
```diff
@@ -0,0 +1,64 @@
+export const usePaymentRequestEnabledSettings = () => {
+	const { updateIsPaymentRequestEnabled } = useDispatch( STORE_NAME );
+
+	return useSelect( ( select ) => {
+		const { getIsPaymentRequestEnabled } = select( STORE_NAME );
+
+		return [ getIsPaymentRequestEnabled(), updateIsPaymentRequestEnabled ];
+	} );
```
I recommend separating the value returned by the selector from the value returned by the hook.
Otherwise the selector is returning a new array reference (which it doesn't need to do) and it makes harder to memoize the returned value.
The `useSelect` doesn't need to know about a change on `updateIsPaymentRequestEnabled`, this way.

Just like in `react-redux`, it looks like `useSelect` "remembers" the value returned by the callback function based on the dependencies. So if the dependencies didn't change and the store didn't change, it can just return a "cached" value: https://github.com/WordPress/gutenberg/blob/f242c91b21bbbccf0a0129e4946aadfa4d04baca/packages/data/src/components/use-select/index.js#L109

```suggestion
	const isPaymentRequestEnabled = useSelect( ( select ) => {
		const { getIsPaymentRequestEnabled } = select( STORE_NAME );

		return getIsPaymentRequestEnabled();
	}, [] );

	return [ isPaymentRequestEnabled, updateIsPaymentRequestEnabled ];
```

waclawjacek replied:
Fantastic! :rocket:

frosso commented on client/data/settings/selectors.js:
```diff
@@ -0,0 +1,28 @@
+	return getSettingsState( state ).isSaving || false;
+};
+
+export const getIsPaymentRequestEnabled = ( state ) => {
+	return getSettings( state ).is_payment_request_enabled || false;
+};
```
I understand that we used the same approach on WCPay - however: :D 

Most of these selectors are only used in one instance in the codebase (in the hooks).

What are your thoughts on mostly leveraging the `getSettings()` selector in the hooks?
This would shave some weight on the bundle size (since we wouldn't need to bundle these additional functions (that cannot be minified) and we could maintain a little less code).

So in the hooks, we could just do (as an example):
```js
export const usePaymentRequestEnabledSettings = () => {
	const { updateIsPaymentRequestEnabled } = useDispatch( STORE_NAME );

	const isPaymentRequestEnabled = useSelect( ( select ) => {
		const { getSettings } = select( STORE_NAME );

		return getSettings().is_payment_request_enabled || false;
	}, [] );

	return [ isPaymentRequestEnabled, updateIsPaymentRequestEnabled ];
};
```

waclawjacek replied:
Yes! This will be a departure from what we do in WCPay but as you said, the only client code for these functions are the hooks.

frosso commented on client/data/settings/hooks.js:
```diff
@@ -0,0 +1,64 @@
+export const useSettings = () => {
+	const { saveSettings } = useDispatch( STORE_NAME );
+
+	return useSelect(
+		( select ) => {
+			const {
+				getSettings,
+				hasFinishedResolution,
+				isResolving,
+				isSavingSettings,
+			} = select( STORE_NAME );
+
+			const isLoading =
+				isResolving( 'getSettings' ) ||
+				! hasFinishedResolution( 'getSettings' );
+
+			return {
+				settings: getSettings(),
+				isLoading,
+				saveSettings,
+				isSaving: isSavingSettings(),
+			};
+		},
+		[ saveSettings ]
+	);
```
Same as below
```suggestion
	const settings = useSelect( ( select ) => {
		const { getSettings } = select( STORE_NAME );

		return getSettings();
	}, [] );

	const isLoading = useSelect( ( select ) => {
		const { hasFinishedResolution, isResolving } = select( STORE_NAME );

		return (
			isResolving( 'getSettings' ) ||
			! hasFinishedResolution( 'getSettings' )
		);
	}, [] );

	const isSaving = useSelect( ( select ) => {
		const { isSavingSettings } = select( STORE_NAME );

		return isSavingSettings();
	}, [] );

	return { settings, isLoading, isSaving, saveSettings };
};
```

waclawjacek replied:
That's brilliant! I give this change 5/5 rockets: :rocket: :rocket: :rocket: :rocket: :rocket:

frosso commented on client/data/settings/actions.js:
```diff
@@ -0,0 +1,68 @@
+			data: settings,
+		} );
+
+		yield dispatch( 'core/notices' ).createSuccessNotice(
+			__( 'Settings saved.', 'woocommerce-gateway-stripe' )
+		);
```
(minor) Displaying a notice here caused a few quirks in WCPay. For example, we couldn't display a different messaging when the settings were saved (https://d.pr/i/2MGNwK) and it also displayed a notice when we didn't want to, on page navigation ( https://github.com/Automattic/woocommerce-payments/issues/2005 ).

What do you think about moving the notice logic somewhere else? Maybe in a separate component or via handlers on this function 🤷

waclawjacek replied:
Extracted to a separate issue: #1840

frosso reviewed (commented):
Thanks for the notes in the description! With the newer release we will start supporting only L-2, so hopefully that will be easier to account for in the future.

This seems to be working well! I have a few comments on the implementation.
What seems to be jumping out more at me is the duplicated instantiation of the Stripe gateway and the fact that we're not using the UPE one.

Commit:
Update client/settings/payment-request-section/__tests__/index.test.js

Co-authored-by: Francesco <frosso@users.noreply.github.com>

Commit:
Replace creator function with direct value usage

Props frosso.

Commit:
Get gateway from WC() instead of instantiating

To prevent duplicate hooks from being registered due to object
instantiation, access the existing gateway using
WC()->payment_gateways() instead.

Commit:
Update client/data/settings/hooks.js

Co-authored-by: Francesco <frosso@users.noreply.github.com>

Commit:
Move selector code directly into hook

Props frosso.

Commit:
Update client/data/settings/hooks.js

Co-authored-by: Francesco <frosso@users.noreply.github.com>

Commit:
Fix typo

Commit:
Merge branch 'develop' into add/1717-stripe-settings-persistence

frosso reviewed (approved):
Thanks for the changes! Let's merge this so that we can move forward with the rest of the work 👍

Commit:
Merge branch 'develop' into add/1717-stripe-settings-persistence

frosso commented:
With this branch checked out, I'm now seeing a blank page at http://localhost:8082/wp-admin/admin.php?page=wc-settings&tab=checkout&section=stripe . On `develop` it's working fine.
![Screen Shot 2021-09-13 at 11 56 04 AM](https://user-images.githubusercontent.com/273592/133125482-d6188463-5749-46ba-8921-baf5ba50a15a.png)

Commit:
Fix settings store tests to reflect data extraction in hooks

Also remove commented out test code.

Commit:
Merge branch 'add/1717-stripe-settings-persistence' of github.com:woocommerce/woocommerce-gateway-stripe into add/1717-stripe-settings-persistence

waclawjacek commented:
Thanks for the heads-up! Investigating now.

Commit:
Fix REST controller type declaration

waclawjacek commented:
Fixed. The issue was caused by the REST controller's constructor expecting the traditional gateway and not accepting the UPE version.

frosso commented:
Great, thanks for the quick fix!
